File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string& s) {
        int maxLen = 0, counter[128] = {};
        for (int left = 0, right = 0; right < s.size(); ++right) {
            while (counter[s[right]]) --counter[s[left++]];
            ++counter[s[right]];
            maxLen = max(maxLen, right - left + 1);
        }
        return maxLen;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = counter = maxLen = 0
        for right in range(len(s)):
            while counter & (1 << ord(s[right])):
                counter ^= (1 << ord(s[left]))
                left += 1
            counter ^= (1 << ord(s[right]))
            maxLen = max(maxLen, right - left + 1)
        return maxLen
            